README.md
<h2 align="center"> Flutter widget to create focused menu easily 🚀 </h2>

[![codecov](https://codecov.io/gh/retroportalstudio/focused_menu/branch/master/graph/badge.svg)](https://codecov.io/gh/retroportalstudio/focused_menu) <a href="https://pub.dev/packages/focused_menu"><img src="https://img.shields.io/pub/v/focused_menu.svg" alt="Pub"></a> <a href="https://opensource.org/licenses/MIT"><img src="https://img.shields.io/badge/license-MIT-blue.svg" alt="License: MIT"></a> <a href="https://github.com/retroportalstudio/focused_menu"><img src="https://img.shields.io/github/stars/retroportalstudio/focused_menu?style=social" alt="Pub"></a>


<a href="https://pub.dev/packages/focused_menu/score"><img src="https://img.shields.io/pub/likes/focused_menu?logo=flutter" alt="Pub likes"></a> <a href="https://pub.dev/packages/focused_menu/score"><img src="https://img.shields.io/pub/popularity/focused_menu?logo=flutter" alt="Pub popularity"></a> <a href="https://pub.dev/packages/focused_menu/score"><img src="https://img.shields.io/pub/points/focused_menu?logo=flutter" alt="Pub points"></a>

---

This is an easy to implement package for adding Focused Menu to Flutter Applications


<p float="center">
  <img src="https://github.com/retroportalstudio/focused_menu/blob/master/example/repo_files/focused_menu.gif?raw=true" width="400" />
  <img src="https://github.com/retroportalstudio/focused_menu/blob/master/example/repo_files/using-controller.gif?raw=true" width="400" /> 
  <img src="https://github.com/retroportalstudio/focused_menu/blob/master/example/repo_files/bottomToolbar.gif?raw=true" width="400" />
</p>

---

## Getting Started
Fist install the dependency

### Add dependency

```yaml
dependencies:
  focused_menu: CURRENT_VERSION
```

### Import package

```dart
import 'package:focused_menu/focused_menu.dart';
```
## Usage
To Use, simply Wrap the Widget you want to add Focused Menu to, with FocusedMenuHolder:
```dart
FocusedMenuHolder(
  menuItems: <FocusedMenuItem>[
    FocusedMenuItem(
      title: Text("This is a button"),
      trailing: Icon(Icons.open_in_new),
      onPressed: () {},
    ),
  ],
  child: CircleAvatar(
    child: Image.asset("assets/images/dp_default.png"),
  ),
),
```

---

## Customizations

In order to customize your focused menu you can change any of the available attributes

```dart
FocusedMenuHolder(
  menuOffset: 0,
  controller: _avatarController,
  animateMenuItems: true,
  blurBackgroundColor: Colors.pink,
  blurSize: 20,
  bottomOffsetHeight: 20,
  duration: Duration(milliseconds: 500),
  menuBoxDecoration:
      BoxDecoration(borderRadius: BorderRadius.circular(20)),
  menuItemExtent: 60,
  menuWidth: 200,
  openWithTap: true,
  enableMenuScroll: false,
  toolbarButtons: [
    ToolbarButtonItem(buttonIcon: Icon(Icons.delete,), onPressed: () {}, buttonIconColor: Colors.red),
    ToolbarButtonItem(buttonIcon: Icon(Icons.share), onPressed: () {}, buttonIconColor: Colors.blue),
  ],
  onOpened: () => print('Opened'),
  onClosed: () => print('onClosed'),
  onPressed: () {},
  menuItems: <FocusedMenuItem>[
    FocusedMenuItem(
      backgroundColor: Colors.green,
      title: Text("This is a button"),
      trailing: Icon(Icons.open_in_new),
      onPressed: () {},
    ),
  ],
  child: CircleAvatar(
    child: Image.asset("assets/images/dp_default.png"),
  ),
),
```

<br>
<div align="center" >
  <p>Thanks to all contributors</p>
  <a href="https://github.com/retroportalstudio/focused_menu/graphs/contributors">
    <img src="https://contrib.rocks/image?repo=retroportalstudio/focused_menu" />
  </a>
</div>
<br>


## License
[MIT](https://choosealicense.com/licenses/mit/)


CHANGELOG.md
## [1.0.0] - June 18, 2020

* Add Focused Menu to Any Widget you Want
* Customizations to change The Focused Menu and Animations according to your Application Needs.

## [1.0.1] - June 24,2020

* Added Customizable offset of Menu from Selected Item
* Added BottomOffsetHeight for Screens with Bottom Navigation Bar

## [1.0.2] - December 23,2020

* Added openWithTap: true|false to Open Focused Menu on Tap rather than a Long Press

## [2.0.0] - YYYY-MM-DD

### Added
- Added bottom toolbar.
- Added controller with open/close methods.
- Ability to disable menu scroll.
- Added onOpened and onClosed callbacks.
- PR template added.
- Upgraded app and migrated to null safety.

### Changed
- Text button changed to Icon in various places.
- Refactored: changed files structure for better readability and a single import point.
- Refactored: updated example to match file changes.
- Refactored: made onPressed optional.
- Refactored: added types to methods and changed `isOpened` to private.
- Refactored: changed `trailingIcon` to `trailing` and made it customizable.
- Refactored: changed `toolbarButtons` to `toolbarActions` and now allows a list of widgets.

### Docs
- Improved readme.md.
- Added more examples and updated example images.
- Added callbacks information to readme.

example/lib/example-position.dart
import 'package:flutter/material.dart';
import 'package:focused_menu/focused_menu.dart';

class ExamplePosition extends StatelessWidget {
  const ExamplePosition({Key? key}) : super(key: key);

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(title: const Text('Example Position')),
      body: SafeArea(
          child: SingleChildScrollView(
        padding: const EdgeInsets.all(15.0),
        child: Column(
          children: [
            Padding(
              padding: const EdgeInsets.symmetric(vertical: 16.0),
              child: Text(
                'Tap on the card to open the menu:',
                style: TextStyle(
                  fontSize: 18,
                ),
              ),
            ),

            /// Bottom left (default)
            MenuItemContainer(
              text: 'Bottom left',
              description: 'default',
              backgroundColor: Colors.blue[600],
            ),

            SizedBox(height: 14),

            /// Bottom right
            MenuItemContainer(
              text: 'Bottom right',
              align: Alignment.bottomRight,
              backgroundColor: Colors.blue[700],
            ),

            SizedBox(height: 14),

            /// Bottom center
            MenuItemContainer(
              text: 'Bottom center',
              align: Alignment.bottomCenter,
              backgroundColor: Colors.blue[600],
            ),

            SizedBox(height: 14),

            /// Top left
            MenuItemContainer(
              text: 'Top left',
              description:
                  'The menu should be opened to the top left. If the menu cannot be fully displayed, it will be opened to the bottom left.',
              align: Alignment.topLeft,
              backgroundColor: Colors.deepPurple[500],
            ),

            SizedBox(height: 14),

            /// Top right
            MenuItemContainer(
              text: 'Top right',
              description:
                  'The menu should be opened to the top right. If the menu cannot be fully displayed, it will be opened to the bottom right.',
              align: Alignment.topRight,
              backgroundColor: Colors.deepPurple[700],
            ),

            SizedBox(height: 14),

            /// Bottom center
            MenuItemContainer(
              text: 'Top center',
              align: Alignment.topCenter,
              backgroundColor: Colors.deepPurple[900],
            ),

            SizedBox(height: 14),

            /// Bottom left - end of the screen
            MenuItemContainer(
              text: 'Bottom left - end of the screen',
              description:
                  'The menu should be opened to the bottom left, but because we are at the end of the screen, it will be opened to the top left.',
              align: Alignment.bottomLeft,
              backgroundColor: Colors.orange[600],
            ),

            SizedBox(height: 14),
          ],
        ),
      )),
    );
  }
}

class MenuItemContainer extends StatelessWidget {
  final String text;
  final String? description;
  final Alignment? align;
  final Color? backgroundColor;
  const MenuItemContainer({
    Key? key,
    required this.text,
    this.description,
    this.align,
    this.backgroundColor,
  }) : super(key: key);

  @override
  Widget build(BuildContext context) {
    return FocusedMenuHolder(
      openWithTap: true,
      menuWidth: MediaQuery.of(context).size.width * 0.50,
      blurSize: 5.0,
      menuItemExtent: 45,
      menuOffset: 8,
      menuBoxDecoration: BoxDecoration(
        color: Colors.grey,
        borderRadius: BorderRadius.circular(14),
      ),
      align: align,
      menuItems: [
        FocusedMenuItem(
          title: Text("Open"),
          trailing: Icon(Icons.open_in_new),
          onPressed: () {},
        ),
        FocusedMenuItem(
          title: Text("Share"),
          trailing: Icon(Icons.share),
          onPressed: () {},
        ),
        FocusedMenuItem(
          title: Text("Favorite"),
          trailing: Icon(Icons.favorite_border),
          onPressed: () {},
        ),
        FocusedMenuItem(
          title: Text("Delete", style: TextStyle(color: Colors.redAccent)),
          trailing: Icon(
            Icons.delete,
            color: Colors.redAccent,
          ),
          onPressed: () {},
        ),
      ],
      child: Stack(
        children: [
          AspectRatio(
            aspectRatio: 16 / 9,
            child: Container(
              decoration: BoxDecoration(
                borderRadius: BorderRadius.circular(14),
                color: backgroundColor,
              ),
              clipBehavior: Clip.antiAlias,
              // child: Image.asset('assets/images/image_1.jpg'),
              child: Center(
                child: Column(
                  mainAxisAlignment: MainAxisAlignment.center,
                  children: [
                    Text(
                      text,
                      style: TextStyle(
                        color: Colors.white,
                        fontSize: 24,
                      ),
                    ),
                    if (description != null)
                      Padding(
                        padding:
                            const EdgeInsets.only(top: 12.0, left: 9, right: 9),
                        child: Text(
                          description!,
                          textAlign: TextAlign.center,
                          style: TextStyle(
                            color: Colors.white,
                            fontSize: 14,
                          ),
                        ),
                      ),
                  ],
                ),
              ),
            ),
          ),
        ],
      ),
    );
  }
}
